src/expire.c

```diff
@@ -521,6 +569,45 @@ void expireGenericCommand(client *c, long long basetime, int unit) {
         return;
     }
 
+    /* NX option is set, check current expiry */
+    if (flag & FLAG_NX) {
+        current_expire = getExpire(c->db, key);
+        if (current_expire != -1) {
+            addReply(c,shared.czero);
+            return;
+        }
+    }
+
+    /* XX option is set, check current expiry */
+    if (flag & FLAG_XX) {
+        current_expire = getExpire(c->db, key);
+        if (current_expire == -1) {
+            /* reply 0 when the key has no expiry */
+            addReply(c,shared.czero);
+            return;
+        }
+    }
+
+    /* GT option is set, check current expiry */
+    if (flag & FLAG_GT) {
+        current_expire = getExpire(c->db, key);
+        if (when <= current_expire || current_expire == -1) {
+            /* reply 0 when the new expiry is not greater than current */
+            addReply(c,shared.czero);
+            return;
+        }
+    }
+
+    /* LT option is set, check current expiry */
+    if (flag & FLAG_LT) {
+        current_expire = getExpire(c->db, key);
+        if (when >= current_expire) {
```

Comment:
so here, like in the GT case, we wanna consider non-volatile keys as having infinite expire time, right? so any expiration time would be less than infinite, in that case we probably wanna add an ` && current_expire != -1`.

I suppose your reasoning for not adding it was that `when` is surely positive (unix time), so if `current_expire` is -1, there's no need to check (any positive `when` is grater than `-1`, and thus should be aborted since it's not less than infinite).
But IIRC, in theory `when` can be negative at this stage, so what would happen is that instead of proceeding and deleting the non-volatile key after the `checkAlreadyExpired` check, we'll abort here .

let's add a test for that case.

Reply:
Good catch

Reply:
test case added